src/core/auth/wallet-authenticator.ts

```diff
@@ -0,0 +1,342 @@
+  LoginOptionsSchema,
+  VerifyOptionsSchema,
+  VerifyOptions,
+  AuthenticationOptionsSchema,
+} from "../../schema/auth";
+import { RPCConnectionHandler } from "../classes/rpc-connection-handler";
+import { NetworkOrSignerOrProvider } from "../types";
+import { UserWallet } from "../wallet";
+
+/**
+ * Wallet Authenticator
+ * @remarks The wallet authenticator enables server-side applications to securely identify the
+ * connected wallet address of users on the client-side, and also enables users to authenticate
+ * to any backend using just their wallet. It implements the JSON Web Token (JWT) authentication
+ * standard.
+ *
+ * @example
+ * ```javascript
+ * // We specify the domain of the application to authenticate to
+ * const domain = "thirdweb.com"
+ *
+ * // On the client side, we can generate a payload for the connected wallet to login
+ * const loginPayload = await sdk.auth.login(domain);
+ *
+ * // Then on the server side, we can securely verify the connected client-side address
+ * const address = sdk.auth.verify(domain, loginPayload);
+ *
+ * // And we can also generate an authentication token to send to the client
+ * const token = sdk.auth.generate(domain, loginPayload);
+ *
+ * // Finally, the token can be send from the client to the server to make authenticated requests
+ * // And the server can use the following function to authenticate a token and verify the associated address
+ * const address = sdk.auth.authenticate(domain, token);
+ * ```
+ * @beta
+ */
+export class WalletAuthenticator extends RPCConnectionHandler {
+  private wallet: UserWallet;
+
+  constructor(
+    network: NetworkOrSignerOrProvider,
+    wallet: UserWallet,
+    options: SDKOptions,
+  ) {
+    super(network, options);
+    this.wallet = wallet;
+  }
+
+  /**
+   * Login With Connected Wallet
+   * @remarks Client-side function that allows the connected wallet to login to a server-side application.
+   * Generates a login payload that can be sent to the server-side for verification or authentication.
+   *
+   * @param domain - The domain of the server-side application to login to
+   * @param options - Optional configuration options for the login request
+   * @returns Login payload that can be used on the server-side to verify the login request or authenticate
+   *
+   * @example
+   * ```javascript
+   * // Add the domain of the application users will login to, this will be used throughout the login process
+   * const domain = "thirdweb.com";
+   * // Generate a signed login payload for the connected wallet to authenticate with
+   * const loginPayload = await sdk.auth.login(domain);
+   * ```
+   */
+  public async login(
+    domain: string,
+    options?: LoginOptions,
+  ): Promise<LoginPayload> {
+    const parsedOptions = LoginOptionsSchema.parse(options);
+
+    const signerAddress = await this.wallet.getAddress();
+    const expirationTime =
+      parsedOptions?.expirationTime || new Date(Date.now() + 1000 * 60 * 5);
```

Comment:
5 minutes by default? is that from the standard?

btw you could put this in the zod schema if you want to

Reply:
Originally had the expiration times defaulted in the zod schema, but it was giving me an odd error where it would keep generating the same default time (which wasn't actually correct 5 minutes from now) and was causing errors until I removed it from schema, not sure what the issue was there but this seemed like an even trade.

Reply:
As far as the timings go (5 minutes and the 5 hours one below), these are important security considerations, and it's okay to have them within a reasonable range. The signature is a one time use payload, ie you really only should need to use each signature once to prove your identity, and 5 minutes is a standard lifetime for a one time use/short-lived  token (along with 1 min, 2 mins, etc. - JWT access/refresh tokens usually fall in this range as well).

As far as the 5 hour one, that's another commonly used timing for session times/JWT token times (that don't rely on refresh tokens)

We definitely _never_ want to have an infinite expiration access token - it's a huge security concern because if someone steals your access token, they can hack your account for the remainder of the time that the token is valid, so adding expiration times mitigates this risk. If there is no expiration, the attacker can auth to your account forever, unless the server-side blocks the specific payload UUID or you switch account 😢 - so we definitely want to keep expiration defaulted on the JWT token.